### PR TITLE
Implement transaction queue

### PR DESCRIPTION
**Pull request by plafer**
Closes: #40

Builds on #39.

Left to do:
+ [ ] clean up code (and adhere to coding guidelines)
+ [ ] write tests
+ [ ] document current transaction queue requirements
+ [ ] write docstrings

### CODE REVIEW

File: block-producer/src/tasks/txqueue/mod.rs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        }
+
+        locked_ready_queue.push(proven_tx);
+
+        if locked_ready_queue.len() >= self.options.batch_size {
+            // FIXME: Dropping the ready queue here means that 2 tasks could
+            // send a batch, where the first batch will contain `batch_size`
+            // transactions, and the second would contain only 1 transaction.
+            // This is low-risk and has little-to-no impact if it occurs.
+            drop(locked_ready_queue);
```
Comment (bobbinth) [inline block-producer/src/tasks/txqueue/mod.rs:177]: Why do we need to drop the locked queue here? Would we be able to pass it into `send_batch()` and avoid the potential issue described in the comment?

**Review by bobbinth — COMMENTED** [review]
Thank you! Looks good! This is not a full review yet. I left some comments inline - mostly nits and questions.

File: block-producer/src/tasks/txqueue/mod.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
+//! The transaction queue takes transactions coming in, validates them, and
+//! eventually sends them out in a batch. We say "sending a batch" to represent
+//! handing over a set of transactions to the batch builder.
```
[the remark this answers is not in the document]
Comment (plafer) [reply block-producer/src/tasks/txqueue/mod.rs]: fixed!

File: block-producer/src/tasks/txqueue/mod.rs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    async fn on_read_transaction(
+        self: Arc<TxQueue<HandleIn, HandleOut>>,
+        proven_tx: ProvenTransaction,
+        timer_task_handle: TimerTaskHandle,
+    ) {
```
[the remark this answers is not in the document]
Comment (plafer) [reply block-producer/src/tasks/txqueue/mod.rs]: Agreed, I prefer `on_transaction()` too. Fixed!

File: block-producer/src/tasks/txqueue/mod.rs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    type ReadTxError: Debug;
+
+    async fn read_transaction(&self) -> Result<ProvenTransaction, Self::ReadTxError>;
+}
+
+/// Contains all the methods for the transaction queue to send messages out.
+#[async_trait]
+pub trait TxQueueHandleOut: Send + Sync + 'static {
+    type VerifyTxError: Debug;
+    type TxVerificationFailureReason: Debug + Send;
```
[the remark this answers is not in the document]
Comment (plafer) [reply block-producer/src/tasks/txqueue/mod.rs]: `VerifyTxError` is meant to capture all "unexpected errors" that (probably) make the system halt (e.g. the underlying channel closed). `TxVerificationFailureReason` represents all "expected issues", i.e. actual verification errors; e.g. the transaction is already in the queue.

I documented this in the error's docstring.

`sled.rs`'s [great blog post](https://sled.rs/errors.html) goes into great detail about the benefits of this pattern.

File: block-producer/src/tasks/txqueue/mod.rs (hunk 1 above)
Comment (plafer) [reply block-producer/src/tasks/txqueue/mod.rs]: I was concerned with potential future deadlocks that arise by holding a mutex across a `tokio::spawn()`. I fixed it differently: `send_batch()`'s caller is responsible for draining the queue.

Fixed in [99542df](https://github.com/0xPolygonMiden/miden-node/pull/44/commits/99542df8374b4d255740e6f572a4e7d4a8a9a7ae)

**Comment by plafer** [comment]
The rework changed the interface of how tasks communicate. The previous one turned out to have 2 important shortcomings:
1. it was not possible to implement a `getX() -> X` message, where a task would initiate a request for some data to another task. All data needed to be pushed out of tasks.
2. It forced us to use a pair of channels to implement the communication, which is very bad because it forces all calls to a given method to be serialized.

The new interface solves both of these issues! Communication between tasks is modeled as a set of requests/responses, and hence I called it `RpcClient` and `RpcServer` (we should find a name other than "rpc" though). The above problems are solved as:

1. to implement a `getX() -> X` relationship, create a client/server pair where the request is `()` and response is `X`.
2. the solution to allow parallelization of many calls to the same method is, as the server, to receive requests on a long-lived channel *as well as a oneshot channel end to send the response on*. When the response is ready, we use that channel. This is implemented (quite succinctly!) [here](https://github.com/0xPolygonMiden/miden-node/blob/9b9e4f01a69b09a12afe4e16e2e7be08b449c2d8/block-producer/src/rpc/mod.rs#L77-L84).

Note that as implemented, if `tx1` comes in before `tx2`, `tx2` could still be added to the internal queue before `tx1` in the case where `tx2`'s verification was faster. There are no requirements on this behavior as of yet though, so I left it as is.

File: block-producer/src/tasks/txqueue/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply block-producer/src/tasks/txqueue/mod.rs]: Interesting post. To me the question is why propagate the unrecoverable error up the stack? If the goal of that error is to abort why not just `panic!()`?

(It seems the types no longer exist, I'm asking more out of curiosity to understand the rationale).

**Review by hackaugusto — COMMENTED** [review]
If I understood the code correctly, this has a generic interface to implement the block producer and txpool. The current implementation is communicating over a mpsc channel for testing purposes, and later on these interfaces will be replaced with gRPC implementations. Let me know if I got it right

File: block-producer/src/tasks/txqueue/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (plafer) [reply block-producer/src/tasks/txqueue/mod.rs]: Good question. Because `panic!()`ing right away isn't always what you want to do. Maybe you want to allow the caller (and every other task in the system) to log some things before ultimately failing. Or maybe you want to try to recover: e.g. if you're talking to another task which panics, you could try to restart that task, and try again.

TLDR: when you want to do anything more sophisticated than outright crashing

File: block-producer/src/lib.rs [hunk 5]
```diff
@@ -1,2 +1,12 @@
...
+use miden_objects::transaction::ProvenTransaction;
+
 #[cfg(test)]
 pub mod test_utils;
+
+// TODO: Better name than rpc?
+pub mod rpc;
+pub mod tasks;
+
+type SharedProvenTx = Arc<ProvenTransaction>;
```
[the remark this answers is not in the document]
Comment (plafer) [reply block-producer/src/lib.rs:12]: Indeed, I will add better docs everywhere.

File: block-producer/src/rpc/mod.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+pub struct RpcServer<Request, Response, S>
+where
+    Request: Send + 'static,
+    Response: Send + 'static,
+    S: ServerImpl<Request, Response>,
+{
+    recv_requests: UnboundedReceiver<(Request, oneshot::Sender<Response>)>,
+    server_impl: Arc<S>,
+}
```
[the remark this answers is not in the document]
Comment (plafer) [reply block-producer/src/rpc/mod.rs:67]: It is necessary, because this is how all tasks within the txpool/block producer will communicate. For any message that needs to be sent between 2 tasks, an `RpcClient`/`RpcServer` pair will be used. And three things can vary:
1. the request type
2. the response type
3. how the server turns the request into a response

Hence the 3 generic types

**Comment by plafer** [comment]
> The current implementation is communicating over a mpsc channel for testing purposes, and later on these interfaces will be replaced with gRPC implementations.

Actually we were thinking that perhaps the `TxQueue`, `BatchBuilder`, `BlockBuilder` and `StateView` would always live on the same machine, and each would take care of distributing their work to other machines as an implementation detail. But I guess that's TBD, and for now it's definitely enough to have them all run on the same machine.

**Comment by plafer** [comment]
Suggestion for better names of current "rpc" terminology:

- `RpcImpl` -> `MessageHandler`
- `RpcClient` -> `MessageSender`
- `RpcServer` -> `MessageReceiver`

**Comment by plafer** [comment]
Closing in favor of #47
